intake/config.py

```diff
@@ -27,53 +27,77 @@
     'cache_download_progress': True,
     'logging': 'INFO',
     'catalog_path': [],
-    'persist_path': posixpath.join(confdir, 'persisted')
-    }
-conf = {}
+    'persist_path': posixpath.join(confdir, 'persisted'),
+    'package_scan': False
+}
 
 
 def cfile():
     return make_path_posix(
         os.getenv('INTAKE_CONF_FILE', posixpath.join(confdir, 'conf.yaml')))
 
 
-def reset_conf():
-    """Set conf values back to defaults"""
-    conf.clear()
-    conf.update(defaults)
-
-
-def save_conf(fn=None):
-    """Save current configuration to file as YAML
-
-    If not given, uses current config directory, ``confdir``, which can be
-    set by INTAKE_CONF_DIR.
-    """
-    if fn is None:
-        fn = cfile()
-    try:
-        os.makedirs(os.path.dirname(fn))
-    except (OSError, IOError):
-        pass
-    with open(fn, 'w') as f:
-        yaml.dump(conf, f)
-
-
-def load_conf(fn=None):
-    """Update global config from YAML file
-
-    If fn is None, looks in global config directory, which is either defined
-    by the INTAKE_CONF_DIR env-var or is ~/.intake/ .
-    """
-    if fn is None:
-        fn = cfile()
-    if os.path.isfile(fn):
-        with open(fn) as f:
-            try:
-                conf.update(yaml_load(f))
-            except Exception as e:
-                logger.warning('Failure to load config file "{fn}": {e}'
-                               ''.format(fn=fn, e=e))
+class Config(dict):
+    def __init__(self, filename=None, **kwargs):
```

Comment:
A pattern I like sometimes is doing the file-based path in a classmethod, as in:

```
Config.from_file(filename, **kwargs)
```

with the `**kwargs` for optional overrides of what's in the file. This way, the parsing is moved out of the single privileged initializer. Also, as this subclasses `dict`, it avoids breaking compatibility with the superclass init.

Reply:
You are right, that would be nicer. There are two complication
- there always is a filename, taken from defaults if not otherwise given, so if `__init__` isn't to handle this, then *every* instantiation would go via `from_file`
- that self.filename also appears in save(). But maybe it shouldn't?